Appraisals

```diff
@@ -127,7 +127,7 @@ if ruby_version?('2.1')
     gem 'semantic_logger', '~> 4.0'
     gem 'sequel', '~> 4.0', '< 4.37'
     gem 'shoryuken'
-    gem 'sidekiq', '~> 3.5.4'
+    gem 'sidekiq', '~> 4.0.0'
```

Comment:
Our test suite runs [for all major Sidekiq versions, from 3.5 to 6.2](https://grep.app/search?q=sidekiq%20%5C%28%5Cd&regexp=true&filter[repo][0]=DataDog/dd-trace-rb&filter[path][0]=gemfiles/).

I noticed that you needed to bump up only Ruby 2.1, which uses Sidekiq 3.5.4.

What's the impact of leaving this 3.5.4 support for Sidekiq? Is it possible to keep support at 3.5.4, while dynamically skipping new features introduced in this PR if necessary?

Reply:
We can definitely do that! Out of curiosity, is there a particular rule of thumb DataDog follows for backwards compatibility? That'd be great to know for any future contributions.

Reply:
We always try to not break existing behaviour: if it was working in version X, should work in version X+1.

If we want to improve things, but these improvements only work with newer versions of the integrated library (Sidekiq in our case), it's reasonable to have separate code paths: one for old behaviour, another for new and improved behaviour. We normally only do this for brand new features, not for two versions of an existing feature (unless strictly technically required).

So, for example, adding a few additional spans or additional tags for only more recent versions of Sidekiq is ok. But having, for example, the `span.resource` for the main Sidekiq span changing depending on Sidekiq version is not great, as people will upgrade Sidekiq and all of the sudden their instrumentation has changed as well, breaking dashboards and monitors.